src/content/learn/typescript.md

```diff
@@ -0,0 +1,474 @@
+
+```ts
+const initialState = { count: 0 };
+
+export default function App() {
+  const [state, dispatch] = useReducer<State>(stateReducer, initialState);
+}
+```
+
+### `useContext` {/*typing-usecontext*/}
+
+The [`useContext` hook](/reference/react/useContext) is a technique for passing data down the component tree without having to pass props through components. It is used by creating a provider component and often by creating a hook to consume the value in a child component.
+
+The type of the value provided by the context is inferred from the value passed to the `createContext` call:
+
+<Sandpack>
+
+```tsx App.tsx active
+import { createContext, useContext, useState } from 'react';
+
+type Theme = "light" | "dark" | "system";
+const ThemeContext = createContext<Theme>("system");
+
+const useGetTheme = () => useContext(ThemeContext);
+
+export default function MyApp() {
+  const [theme, setTheme] = useState<Theme>('light');
+
+  return (
+    <ThemeContext.Provider value={theme}>
+      <MyComponent />
+    </ThemeContext.Provider>
+  )
+}
+
+function MyComponent() {
+  const theme = useGetTheme();
+
+  return (
+    <div>
+      <p>Current theme: {theme}</p>
+    </div>
+  )
+}
+```
+
+```js App.js hidden
+import AppTSX from "./App.tsx";
+export default App = AppTSX;
+```
+
+</Sandpack>
+
+This technique works when you have an default value which makes sense - but there are occasionally cases when you do not, and in those cases `null` can feel reasonable as a default value. However, to allow the type-system to understand your code, you need to explicitly set `ContextShape | null` on the `createContext`. 
+
+This causes the knock-on issue that you need to eliminate the `| null` in the type for context consumers. Our recommendation is to have the hook do a runtime check for it's existence and throw an error when not present:
+
+```js {5, 16-20}
+import { createContext, useContext, useState, useMemo } from 'react';
+
+// This is a simpler example, but you can imagine a more complex object here
+type ComplexObject = {
+  kind: string
+};
+
+// The context is created with `| null` in the type, to accurately reflect the default value.
+const Context = createContext<ComplexObject | null>(null);
+
+// The `| null` will be removed via the check in the hook.
+const useGetComplexObject = () => {
+  const object = useContext(Context);
+  if (!object) { throw new Error("useGetComplexObject must be used within a Provider") }
+  return object;
+}
+
+export default function MyApp() {
+  const object = useMemo(() => ({ kind: "complex" }), []);
+
+  return (
+    <Context.Provider value={object}>
+      <MyComponent />
+    </Context.Provider>
+  )
+}
+
+function MyComponent() {
+  const object = useGetComplexObject();
+
+  return (
+    <div>
+      <p>Current object: {object.kind}</p>
+    </div>
+  )
+}
+```
+
+### `useRef` {/*typing-useref*/}
```

Comment:
Refs and effects are "escape hatches" in the learn docs so let's skip them?

Reply:
I agree, and I think this can be a React team decision which I'll be happy with. I have no pushback on removing `useEffect` as it's really only there because it would be the only one _not_ there.

`useRef` though has this tricky _"either it's mutable, or it's not"_ thing that you kinda have to learn or it doesn't really fit in your head when you first try it - because for using it inline in JSX you need to really have that `|null` . I think beginners get stuck on this bit, and "react typescript useRef" finding this link would describe everything they need to know.

Reply:
It's probably best add to the `useRef` docs in a "TS usage" section or something. At least for the APIs we consider escape hatches.

Reply:
Tbh., it would feel very weird if this was the one hook where the TS instructions would be buried somewhere completely else. In that case, TS usage instrucations would need to be added to pretty much every single API page.

Maybe just keep it here at the bottom of the page under "other hooks"?

Reply:
I don't think this page will be digestable if we add TS usage instructions for every API. Mid-term, React with TypeScript will fit naturally into docs and won't be an after-thought that a single page would imply. This page is really just to get you started.

I'm fine if this is just temporary until we figure out TS in demos but if the intention is to keep TS in a single page, I'd like to discuss this a bit more.

Reply:
I'd *love* to see TS everywhere in the docs, and then no complaints from me - but so far it seemed like a very hard thing to have any form of TS in the docs, so I didn't expect there to be a high chance for that.

Reply:
Moved `useRef` to its reference page while mentioning at the bottom that individual pages may contain TS usage if necessary. I removed the `useEffect` section since it didn't seem like it contained relevant information. 
[`c0e6963` (#6120)](https://github.com/reactjs/react.dev/pull/6120/commits/c0e6963a549fda1b845817d5e61ca8fe11d170b1)